x-pack/auditbeat/module/system/login/utmp.go

```diff
@@ -0,0 +1,571 @@
+			} else {
+				utmpRecords, err = r.readAfter(path, nil)
+			}
+
+			if err != nil {
+				return nil, errors.Wrapf(err, "error reading file %v", path)
+			} else if len(utmpRecords) == 0 {
+				return nil, fmt.Errorf("unexpectedly, there are no new records in file %v", path)
+			} else {
+				for _, utmp := range utmpRecords {
+					loginRecord := r.processLoginRecord(utmp)
+					if loginRecord != nil {
+						loginRecord.Origin = path
+						loginRecords = append(loginRecords, *loginRecord)
+					}
+				}
+			}
+		}
+	}
+
+	return loginRecords, nil
+}
+
+// deleteOldRecords cleans up old file records where the inode no longer exists.
+func (r *UtmpFileReader) deleteOldRecords(fileInfos map[string]os.FileInfo) {
+	for savedInode := range r.fileRecords {
+		found := false
+		for _, fileInfo := range fileInfos {
+			inode := Inode(fileInfo.Sys().(*syscall.Stat_t).Ino)
+			if inode == savedInode {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			r.log.Debugf("Deleting file record for old inode %d.", savedInode)
+			delete(r.fileRecords, savedInode)
+		}
+	}
+}
+
+func (r *UtmpFileReader) fileInfos() (map[string]os.FileInfo, error) {
+	paths, err := filepath.Glob(r.filePattern)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to expand file pattern")
+	}
+
+	// Sort paths in reverse order (oldest/most-rotated file first)
+	sort.Sort(sort.Reverse(sort.StringSlice(paths)))
+
+	fileInfos := make(map[string]os.FileInfo, len(r.fileRecords))
+	for _, path := range paths {
+		fileInfo, err := os.Stat(path)
+		if err != nil {
+			if os.IsNotExist(err) {
+				// Skip - file might have been rotated out
+				r.log.Debugf("File %v does not exist anymore.", path)
+				continue
+			} else {
+				return nil, errors.Wrapf(err, "unexpected error when looking up file %v", path)
+			}
+		} else if fileInfo.Sys() == nil {
+			return nil, fmt.Errorf("empty stat result for file %v", path)
+		}
+
+		if r.log.IsDebug() {
+			inode := Inode(fileInfo.Sys().(*syscall.Stat_t).Ino)
+			if _, isKnownFile := r.fileRecords[inode]; !isKnownFile {
+				r.log.Debugf("Found new file: %v (size=%v)", path, fileInfo.Size())
+			}
+		}
+
+		fileInfos[path] = fileInfo
+	}
+
+	return fileInfos, nil
+}
+
+func (r *UtmpFileReader) updateFileRecord(inode Inode, size int64, utmpRecords *[]Utmp) {
+	newFileRecord := FileRecord{
+		Inode: inode,
+		Size:  size,
+	}
+
+	if utmpRecords != nil && len(*utmpRecords) > 0 {
+		newFileRecord.LastUtmp = (*utmpRecords)[len(*utmpRecords)-1]
+	} else {
+		oldFileRecord, found := r.fileRecords[inode]
+		if found {
+			newFileRecord.LastUtmp = oldFileRecord.LastUtmp
+		}
+	}
+
+	r.fileRecords[inode] = newFileRecord
+}
+
+// ReadAfter reads a UTMP formatted file (usually /var/log/wtmp*)
+// and returns the records after the provided last known record.
+// If record is nil, it returns all records in the file.
+func (r *UtmpFileReader) readAfter(path string, lastKnownRecord *Utmp) ([]Utmp, error) {
```

Comment:
The format of this utmp struct seems pretty simple. I think you could use a pure Go decoder and then not have to deal with reading the full file each time. Currently there's no way to seek to the last offset so you have read the whole file over.

For example: https://github.com/neko-neko/utmpdump/blob/master/utmp/utmp.go

Reply:
Changed to pure Go decoder.